src/test/unit/bqrc_kalman_unittest.cc

```diff
@@ -0,0 +1,197 @@
+
+#include <stdint.h>
+#include <stdio.h>
+#include <time.h>
+#include <math.h>
+#include "gtest/gtest.h"
+
+extern "C" {
+    #include "common/maths.h"
+    #include "common/filter.h"
+}
+
+#define M_PI_FLOAT 3.14159265358979323846f
+#define SAMPLE_SET_SIZE 500000
+#define PASSING_CORRELATION 0.9999f
+
+
+double corr(double *x, double *y, int startIndex, int n)
+{
+    double sumX = 0;
+    double sumY = 0;
+    double sumXY = 0;
+    double squareSumX = 0;
+    double squareSumY = 0;
+
+    for (int i = startIndex; i <= startIndex + n; i++) {
+        sumX += x[i];
+        sumY += y[i];
+        sumXY += x[i] * y[i];
+        squareSumX += pow(x[i], 2);
+        squareSumY += pow(y[i], 2);
+    }
+    return (n * sumXY - sumX * sumY)
+                  / sqrt((n * squareSumX - sumX * sumX)
+                      * (n * squareSumY - sumY * sumY));
+}
+
+double mean(double *x, int startIndex, int n)
+{
+    double varSum = 0;
+
+    for (int i = startIndex; i <= startIndex + n; i++) {
+        varSum += x[i];
+    }
+
+    return varSum / n;
+}
+
+double var(double *x, int startIndex, int n) {
+
+    double varSum = 0;
+    const double mu = mean(x, startIndex, n);
+
+    for (int i = startIndex; i <= startIndex + n; i++) {
+        varSum += pow(x[i] - mu,2);
+    }
+
+    return varSum / n;
+}
+
+
+TEST(BQRCKalmanTest, BQRCKalmanRandomCorrelationTest)
+{
+    // Input parameters exposed to pilot
+    const uint16_t q = 400;
+    const uint16_t r = 88;
+    const uint8_t p = 0;
+    const float rate = 32000;
+
+    fastKalman_t fkfFilter;
+    fastKalmanInit(&fkfFilter, q, r, p);
+
+    const float dT = 1 / rate;
+    const float bqrcQ = fkfFilter.q;
+    const float bqrcR = fkfFilter.r;
+    const float bqrcK = 2 * sqrt(bqrcQ) / (sqrt(bqrcQ + 4 * bqrcR) + sqrt(bqrcQ));
+    const float cutoffFrequency = 1 / (2 * M_PI_FLOAT * 0.5f * dT * (1 / bqrcK - 2));
+
+    biquadFilter_t bqrcFilter;
+    biquadRCFIR2FilterInit(&bqrcFilter, cutoffFrequency, dT);
+
+    double kalmanVec[SAMPLE_SET_SIZE];
+    double bqrcVec[SAMPLE_SET_SIZE];
+    double lastK = 0;
+    int kConvergeSample = 0;
+
+    srand(time(NULL));
+    for (int i = 0; i < SAMPLE_SET_SIZE; i++) {
+        const float randVal = (rand() % 20000) * 1e-1f;
+
+        kalmanVec[i] = fastKalmanUpdate(&fkfFilter, randVal);
+        bqrcVec[i] = biquadFilterApply(&bqrcFilter, randVal);
+
+        if (fkfFilter.k > lastK) {
```

Comment:
`fabs(fkfFilter.k - lastK) > eps` with reasonable eps is cleaner

Reply:
Agreed, it also doesn't give `k` a chance to overshoot and then stabilize, even though that doesn't happen in reality.